pkg/agent/route/route_windows.go

```diff
@@ -213,109 +236,121 @@ func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
+		return nil
 	}
-	// For NodePort Service, a new NetNat for NetNatStaticMapping is needed.
-	err := util.NewNetNat(antreaNatNodePort, virtualNodePortDNATIPv4Net)
+
+	// Enlarge the ClusterIP CIDR with the ClusterIP. Note that, if the ClusterIP CIDR doesn't exist, just generate a CIDR
+	// with the ClusterIP.
+	newClusterIPCIDR, err := util.ExtendCIDRWithIP(curClusterIPCIDR, svcIP)
 	if err != nil {
-		return err
+		return fmt.Errorf("enlarge the destination CIDR with an error: %w", err)
 	}
 
-	return nil
-}
+	// Generate a route with the new ClusterIP CIDR and install it.
+	route := generateRoute(newClusterIPCIDR, gw, linkIndex, metric)
+	if err = util.ReplaceNetRoute(route); err != nil {
+		return fmt.Errorf("failed to install new ClusterIP route: %w", err)
+	}
 
-// TODO: Follow the code style in Linux that maintains one Service CIDR.
-func (c *Client) addServiceRoute(svcIP net.IP) error {
-	obj, found := c.hostRoutes.Load(svcIP.String())
-	svcIPNet := util.NewIPNet(svcIP)
+	// Store the new ClusterIP CIDR and the new generated route to serviceRoutes. Then the calculated route can be restored
+	// when it was deleted since members of serviceRoutes are synchronized periodically.
+	c.clusterIPv4CIDR = newClusterIPCIDR
+	c.serviceRoutes.Store(clusterIPv4CIDRKey, route)
+	klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "route", route, "ClusterIP", svcIP)
 
-	// Route: Service IP -> VirtualServiceIPv4 (169.254.0.253)
-	route := &util.Route{
-		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
-		DestinationSubnet: svcIPNet,
-		GatewayAddress:    config.VirtualServiceIPv4,
-		RouteMetric:       util.MetricHigh,
-	}
-	if found {
-		existingRoute := obj.(*util.Route)
-		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) && existingRoute.RouteMetric == route.RouteMetric {
-			klog.V(2).InfoS("Service route already exists", "DestinationIP", route.DestinationSubnet,
-				"Gateway", route.GatewayAddress, "RouteMetric", route.RouteMetric)
-			return nil
+	// Collect stale routes.
+	routes, err := c.listIPRoutesOnGW()
+	if err != nil {
+		return fmt.Errorf("error listing ip routes: %w", err)
+	}
+
+	var staleRoutes []*util.Route
+	if curClusterIPCIDR != nil {
+		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled since
+		// a new route with a newly calculated destination CIDR has been installed.
+		route.DestinationSubnet = curClusterIPCIDR
+		staleRoutes = []*util.Route{route}
+
+		// TODO: delete the code below in the future. The code below is used to delete stale route for the ClusterIP before
+		//   this commit. For Antrea which is not upgraded from the version before this commit, the code below is unnecessary.
+		for _, rt := range routes {
+			ones, _ := rt.DestinationSubnet.Mask.Size()
+			if ones == ipv4AddrLength && rt.DestinationSubnet.IP.Equal(svcIP) {
+				staleRoutes = append(staleRoutes, rt)
+			}
```

Comment:
why is it needed? isn't curClusterIPCIDR nil when agent restarts then we could remove all unexpected routes by the other branch?

Reply:
The other branch doesn't remove the route whose prefix is 32. Before this commit, every ClusterIP has its own route whose prefix is 32.  For example, we have 3 ClusterIPs, 10.96.0.1, 10.96.1.1 and 10.96.2.1. Before the version of this commit, the routes should be on Windows Node:

- 10.96.0.1/32
- 10.96.1.1/32
- 10.96.2.1/32

After upgrading to the version after this commit, when agent restarting and reconciling these ClusterIPs, the new aggregated route for ClusterIPs will be created, but the stale routes are still there, we need this piece of code to clean them during generating the aggregated route.

Reply:
It's because the code was not generic. Does the code I suggested for linux handles it?

Reply:
The code you suggested for Linux is much better, but it cannot handle the issue on Windows. I can remove the code here. If so, we will get the following route list:
- 10.96.0.1/32 (orphan route)
- 10.96.1.1/32 (orphan route)
- 10.96.2.1/32 (orphan route)
- 10.96.0.0/23 (aggregated from 10.96.0.1, 10.96.1.1 and 10.96.2.1)

The code is used to clean the orphan routes, but it is also ok not to clean them.